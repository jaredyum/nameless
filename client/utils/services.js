/* eslint-disable no-console */
import axios from 'axios';

// Firebase config.
import firebase from 'services/firebase';

// Errors
import { API_ERRORS } from 'copy/Global/errors';

/**
 * Decorates the headers for a request.
 * @param {string} token The firebase token.
 * @return {!Object} The decorated headers object.
 */
const decorateHeaders = token => ({
  Authorization: `Bearer ${token}`,
  'Content-Type': 'application/json'
});

export const defaultCatch = (endPoint, source = 'services') => (e) => {
  console.error(`${source}/${endPoint}`, e);
  throw e;
};

/**
 * Extracts and parses the Firebase token to use for requests.
 * @return {!Promise} A promise resolving with the firebase token.
 */
export const getUserToken = () => firebase
  .then(auth => auth.currentUser)
  .then(user => user.getIdToken())
  .catch(defaultCatch('getUserToken'));

/**
 * Request wrapper for all request types (GET, POST, PUT, DELETE, etc).
 * @param {!Object} config The request configuration.
 * @param {string} config.url The URL for the request.
 * @param {string} config.method The method (GET, POST, etc) for the request.
 * @param {!Object} config.payload Optional request payload.
 * @return {!Promise} Promise containing the axios response.
 */
export const makeRequest = ({
  url,
  method = 'get',
  payload = null
}) => getUserToken()
  .then(token => ({
    headers: decorateHeaders(token),
    method,
    url,
    ...(
      payload ? { data: payload } : {}
    )
  }))
  .then(axios)
  .then((res) => {
    if (res && (res.status >= 200 || res.status < 400)) {
      return res;
    }
    throw new Error(API_ERRORS.BAD_REQUEST);
  })
  .then((res) => {
    if ('location' in res.headers) {
      return makeRequest(res.headers.location, 'GET', null, true);
    }

    return res;
  });
