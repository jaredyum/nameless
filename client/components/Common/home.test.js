import { testCommonComponentAttrs } from 'utils/tests';

import Home from './home';

describe('<Home /> Component', () => {
  testCommonComponentAttrs(Home);
});
