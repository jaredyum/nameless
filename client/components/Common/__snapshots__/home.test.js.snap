// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Component when rendering should render a component matching the snapshot 1`] = `
<WithStyles(ContentSection)>
  <WithStyles(Typography)
    align="center"
    gutterBottom={true}
    variant="display3"
  >
    Welcome to the Home Page!
  </WithStyles(Typography)>
  <WithStyles(Typography)
    align="center"
    gutterBottom={true}
    variant="body1"
  >
    Hold tight, we've got some good stuff on the way ;-)
  </WithStyles(Typography)>
</WithStyles(ContentSection)>
`;
